blockchain/utxocache.go

```diff
@@ -126,3 +133,410 @@ func (ms *mapSlice) deleteMaps() {
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// addTxIn will add the given input to the cache if the previous outpoint the txin
+// is pointing to exists in the utxo set.  The utxo that is being spent by the input
+// will be marked as spent and if the utxo is fresh (meaning that the database on disk
+// never saw it), it will be removed from the cache.
+func (s *utxoCache) addTxIn(txIn *wire.TxIn, stxos *[]SpentTxOut) error {
+	// Ensure the referenced utxo exists in the view.  This should
+	// never happen unless there is a bug is introduced in the code.
+	entries, err := s.fetchEntries([]wire.OutPoint{txIn.PreviousOutPoint})
+	if err != nil {
+		return err
+	}
+	if len(entries) != 1 || entries[0] == nil {
+		return AssertError(fmt.Sprintf("missing input %v",
+			txIn.PreviousOutPoint))
+	}
+
+	// Only create the stxo details if requested.
+	entry := entries[0]
+	if stxos != nil {
+		// Populate the stxo details using the utxo entry.
+		var stxo = SpentTxOut{
+			Amount:     entry.Amount(),
+			PkScript:   entry.PkScript(),
+			Height:     entry.BlockHeight(),
+			IsCoinBase: entry.IsCoinBase(),
+		}
+
+		*stxos = append(*stxos, stxo)
+	}
+
+	// Mark the entry as spent.
+	entry.Spend()
+
+	// If an entry is fresh it indicates that this entry was spent before it could be
+	// flushed to the database. Because of this, we can just delete it from the map of
+	// cached entries.
+	if entry.isFresh() {
+		// If the entry is fresh, we will always have it in the cache.
+		s.cachedEntries.delete(txIn.PreviousOutPoint)
+		s.totalEntryMemory -= entry.memoryUsage()
+	} else {
+		// Can leave the entry to be garbage collected as the only purpose
+		// of this entry now is so that the entry on disk can be deleted.
+		entry = nil
+		s.totalEntryMemory -= entry.memoryUsage()
+	}
+
+	return nil
+}
+
+// addTxIns will add the given inputs of the tx if it's not a coinbase tx and if
+// the previous output that the input is pointing to exists in the utxo set.  The
+// utxo that is being spent by the input will be marked as spent and if the utxo
+// is fresh (meaning that the database on disk never saw it), it will be removed
+// from the cache.
+func (s *utxoCache) addTxIns(tx *btcutil.Tx, stxos *[]SpentTxOut) error {
+	// Coinbase transactions don't have any inputs to spend.
+	if IsCoinBase(tx) {
+		return nil
+	}
+
+	for _, txIn := range tx.MsgTx().TxIn {
+		err := s.addTxIn(txIn, stxos)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// connectTransaction updates the cache by adding all new utxos created by the
+// passed transaction and marking and/or removing all utxos that the transactions
+// spend as spent.  In addition, when the 'stxos' argument is not nil, it will
+// be updated to append an entry for each spent txout.  An error will be returned
+// if the cache and the database does not contain the required utxos.
+func (s *utxoCache) connectTransaction(
+	tx *btcutil.Tx, blockHeight int32, stxos *[]SpentTxOut) error {
```

Comment:
Why have a pointer to the slice passed in which is mutated each time, vs just returning the set of UTXOs spent due to the transaction? Is the goal to reduce memory churn my pre-allocating `stxos` ahead of time? Current usage made the control flow a bit harder to follow IMO.

Reply:
That's a good point. There wasn't much thought that went into this besides just re-using the function prototype for `connectTransaction` that was in the current codebase. I'm assuming that was the intention when previous authors wrote this.